### ISSUE TITLE
Improving bcftools stats performance on large variant calling VCF files

### ISSUE THREAD

**Issue by WimSpee** [bcbio/bcbio-nextgen#3036]
Hi, 

Would it be possible to improve the bcftools stats performance on large variant calling VCF files? 

See the 3 listed options and also the detailed description below and the timestamps at the bottom of this issue opening. 

> For some species we currently have large variant tables that are also getting larger with each sample that we whole genome sequence. 
> 
> For example for one species the current variant table is 666GB big, with 580M variants and with just 176 whole genome sequenced samples.  
> 
> When we whole genome sequence new samples of the same species we first run a alignment and variant calling for these new samples.  That goes fast because since there there are not many samples and the input and output data is small. 
> 
> Then we start a new  bcbio analysis from the old and new existing BAM and GVCF files.  
> Preparing the input and creating a new multi-sample VCF file from the expanded set of GVCF files just takes a few hours to days for each step.  Total is c.a. 2 days, using 20 jobs of  10cpu. 
> 
> The limiting step at this moment is the per sample bcftools stats that is run on the 666GB VCF.gz file. This takes c.a. 20 hours per sample.   With the 176 samples and running 20 jobs in parallel, total time is expected to be 7.5 days.  
> 
> This is almost 4 times the time needed for creating the new mult-sample VCF file. 
>  It would be very nice if the time needed to get the per sample variants stats with bcftools could be decreased. 
> 
> From this previous ticket I know that the creating the variant stats time takes by far the longest of all steps in the QC stage. 
> https://github.com/bcbio/bcbio-nextgen/issues/1556

I can imagine 3 possible ways to improve the performance of getting the per sampel variant stats

1.   Calculate the per sample variant stats on a multi-sample BCF file, reading BCF is c.a. 83% faster than reading VCF.gz . This would require an extra intermediary BCF file being created, or the default multi-sample VCF.gz file becoming a BCF file.  This would also make the soft filtering of SNPs and INDELs much faster. 
2.   Do a single pass split to single sample VCF files, then calculate the variant stats on the single sample VCF files.  This would require the large multi-sample VCF.gz file to be read only once. I am not sure if a single pass split to per sample VCF tool exists. But it would not be to hard to write in cyvcf2.
3.   Do a single pass to get the per sample variant stats from the multi-sample VCF.gz file. This seems to be supported by bcftools and whould be much faster since the large VCF.gz file only needs to be read once.   I am not sure yet if MultiQC can build the same report on this single  stats file or multiple bcftools stats files.  This is something I can test.
```
-s, --samples <list> list of samples for sample stats, "-" to include all samples
```

This would generate a single bcftools stats file with the per sample variant information for all samples.  
```
# PSC, Per-sample counts
# PSC   [2]id   [3]sample       [4]nRefHom      [5]nNonRefHom   [6]nHets        [7]nTransitions [8]nTransversions       [9]nIndels      [10]average depth       [11]nSingletons
```

Please let me know if you think that one of the above options would be a good solution. 
Where I have the knowledge / skills I am happy to test and contribute.

See also the bcbio analysis timestamps below. 

Thank you. 

```
[2019-11-28T09:48Z] Timing: organize samples
[2019-11-28T13:53Z] Timing: alignment preparation
[2019-11-28T20:51Z] ipython: delayed_bam_merge
[2019-11-29T01:09Z] Identified 941 parallel analysis blocks

[2019-11-29T01:25Z] ipython: square_batch_region
[2019-12-01T21:56Z] ipython: concat_variant_files

[2019-12-02T00:39Z]  Timing: variant post-processing

[2019-12-02T00:39Z]  Timing: quality control
[2019-12-02T00:39Z]  QC: 530_01 variants
[2019-12-02T00:39Z]  QC: 530_02 variants
[2019-12-02T00:39Z]  QC: 530_07 variants
[2019-12-02T00:39Z]  QC: 530_04 variants
[2019-12-02T00:39Z]  QC: 530_05 variants
[2019-12-02T00:39Z]  QC: 530_06 variants
[2019-12-02T00:40Z]  QC: 560_01 variants
[2019-12-02T00:40Z]  QC: 574_01 variants
[2019-12-02T00:40Z]  QC: 960_09 variants
[2019-12-02T00:40Z]  QC: 960_10 variants
[2019-12-02T00:42Z]  QC: 032_01 variants
[2019-12-02T00:42Z]  QC: 032_02 variants
[2019-12-02T00:43Z]  QC: 032_03 variants
[2019-12-02T00:43Z]  QC: 032_04 variants
[2019-12-02T00:43Z]  QC: 032_05 variants
[2019-12-02T00:45Z]  QC: 032_06 variants
[2019-12-02T00:45Z]  QC: 032_07 variants
[2019-12-02T00:47Z]  QC: 092_01 variants
[2019-12-02T00:48Z]  QC: 092_04 variants
[2019-12-02T00:53Z]  QC: 294_01 variants
[2019-12-02T20:15Z]  QC: 294_02 variants
[2019-12-02T20:20Z]  QC: 294_03 variants
[2019-12-02T20:25Z]  QC: 294_04 variants
[2019-12-02T20:29Z]  QC: 294_05 variants
[2019-12-02T20:31Z]  QC: 294_06 variants
[2019-12-02T20:35Z]  QC: 519_01 variants
[2019-12-02T20:48Z]  QC: 960_08 variants
[2019-12-02T20:48Z]  QC: 412_01 variants
[2019-12-02T20:48Z]  QC: 368_01 variants
[2019-12-02T20:51Z]  QC: 443_03 variants
[2019-12-02T20:53Z]  QC: 443_04 variants
[2019-12-02T20:56Z]  QC: 443_05 variants
[2019-12-02T21:14Z]  QC: 092_02 variants
[2019-12-02T21:51Z]  QC: 443_02 variants
[2019-12-02T22:08Z]  QC: 442_01 variants
[2019-12-02T22:11Z]  QC: 442_02 variants
[2019-12-02T22:13Z]  QC: 449_01 variants
[2019-12-02T22:20Z]  QC: 449_02 variants
[2019-12-02T22:21Z]  QC: 449_03 variants
[2019-12-02T22:26Z]  QC: 443_01 variants
[2019-12-03T16:09Z]  QC: 146_01 variants
[2019-12-03T16:27Z]  QC: 146_02 variants
```

**Comment by WimSpee** [comment]
I just created a single stats file with the stats of all samples using the following bcftools stats option

```
-s, --samples <list> list of samples for sample stats, "-" to include all samples
```

The resulting stats file does have a block with the per sample variation stats, starting with 
```
# PSC   [2]id   [3]sample       [4]nRefHom      [5]nNonRefHom   [6]nHets        [7]nTransitions [8]nTransversions       [9]nIndels      [10]average depth       [11]nSingletons [12]nHapRef     [13]nHapAlt     [14]nMissing
```

MultiQC however did not recognize the multiple samples in the stats file. 
The reported created by MultiQC just showed the stats for a single sample, with the name of the stats file:
```
DA_1234-gatk-haplotype-joint-annotated
```

Option 3 would thus need additional functionality in MultiQC to detect the stats of the multiple samples in a single bcftools stats file.

**Comment by naumenko-sa** [comment]
Hi @WimSpee !

Thanks for elaborating this use case of large scale population variant calling in bcbio!
We need to think it out, because just using 1) or 2) would add unnecessary overhead for many small-size variant calling projects.

So then 3) might be a solution, or using a separate QC logic for large GVCF joint genotyping projects.

Sergey

**Comment by WimSpee** [comment]
Hi @naumenko-sa  Thank you for your response

**Option 1: Using BCF by default and soft filtering**

I had a look at option 1, to use BCF by default or just for some steps.  The upside of this  would be that also the soft filtering step runs much faster when using BCF as input
`def gatk_snp_cutoff`  and `def gatk_indel_cutoff`
https://github.com/bcbio/bcbio-nextgen/blob/master/bcbio/variation/vfilter.py

But it looks complicated to update bcbio to always or sometimes use BCF instead of VCF.gz.
Just these parts of the code then should be updated or duplicated:
`def _run_genotype_gvcfs_genomicsdb`
https://github.com/bcbio/bcbio-nextgen/blob/ff63c3645a1977ce27c5227ad90cc1d70f083ef3/bcbio/variation/gatkjoint.py#L102
`def convert_to_bcf_and_index`
https://github.com/bcbio/bcbio-nextgen/blob/master/bcbio/variation/vcfutils.py#L590
`_run_concat_variant_files_bcftools`
https://github.com/bcbio/bcbio-nextgen/blob/master/bcbio/variation/vcfutils.py#L471

The soft filtering though is less of an issue, it is less of a bottleneck / scaling problem, and coud be sped up also by doing the soft filtering per chromosome with a naive concat in the end. 

Per chromosome soft filtering is also what I myself implemented outside of bcbio, since starting a bcbio analysis from BAM and GVCF currently does not run soft filtering. See also  https://github.com/bcbio/bcbio-nextgen/issues/2336 

If the current soft filtering approach becomes a bottleneck during analysis of new samples with bcbio, I'll create a ticket. 

**Option 3: Single pass creation of multisample stats file, and feeding into MultiQC**

Back to speeding up the performance of bcftools stats. Option 3 also looks the best to me.  
Maybe it is possible to split the single stats file to multiple stats files that MultiQC can use. 
Or maybe MultiQC can be updated to read in the single multi-sample bcftools stats file. 

I'll first try to split the multi sample stats file and see if the MultiQC created on those makes sense. 
Otherwhise I'll create a ticket at MultiQC for reading in a single bcftools stats file.

**Comment by WimSpee** [comment]
I created a ticket at multiq: https://github.com/ewels/MultiQC/issues/1087

**Comment by WimSpee** [comment]
I did not receive any feedback yet in the MultiQC ticket (might be because of the Christmas/new year holiday). 

For the time being I implemented this 3 step workaround:

**1# Convert the joint vcf.gz file to BCF, and then use the BCF file to create the bcftools stats file per sample. Using BCF is 86% faster than using vcf.gz as input** 
```
# Convert the joint vcf to bcf
bcftools view DA_1550-joint.vcf.gz -Ob -o DA_1550-joint.bcf

# Create a file with all the commands to create the stats file per sample
for sample in `bcftools query -l DA_1550-joint.bcf`; do echo "bcftools stats -s $sample DA_1550-joint.bcf >  ${sample}_bcftools_stats.txt "   ; done > stats_commands.sh

# split this file so that there is a single file with the command per sample
split -l 1 stats_commands.sh  --additional-suffix=.sh stats_commands_

# make the per sample files executable
chmod 0777 stats_commands_*

# submit the per sample stats commands to the cluster
# qig is a internal wrapper of qsub, use whichever cluster submission method that you have
find ./ -name "stats_commands_*sh" -exec qig {} -za \;
```

**2# Put the sample name in the header of each stats file** 

```
# Put the sample name in the header of each stats file 
from os import listdir, getcwd, mkdir
from os.path import isfile, join

cwd = getcwd()

onlyfiles = [f for f in listdir(cwd) if isfile(join(cwd, f))]
only_stats_files = [f for f in onlyfiles if f.endswith("_bcftools_stats.txt")]

mkdir(join(cwd, "fixed"))

out_file = join(cwd, "fixed",  )

for stats_file in only_stats_files:
    name = stats_file.replace("_bcftools_stats.txt", "")
    out_file = join(cwd, "fixed", stats_file)
    with open(stats_file) as in_handle:
         with open(out_file, "w") as out_handle:
             for line in in_handle:
                  if line.startswith("ID\t"):
                      parts = line.split("\t")
                      parts[-1] = "%s\n" % name
                      line = "\t".join(parts)
                  out_handle.write(line)
```


**3# Copy the bcftools stats file to the location where bcbio expects them** 
```
# create the bcbio qc directory for each sample
for sample in `bcftools query -l DA_1550-joint.bcf`; do echo "mkdir -p  ../work/qc/${sample}/variants/"   ; done  > mkdir_commands.sh
sh mkdir_commands.sh

# copy the stats files to the location expected by bcbio
for sample in `bcftools query -l DA_1550-joint.bcf`; do echo "cp  fixed/${sample}_bcftools_stats.txt ../work/qc/${sample}/variants/${sample}_bcftools_stats.txt"   ; done  > copy_commands.sh

sh copy_commands.sh

```

**Comment by naumenko-sa** [comment]
closing for now, will re-open later when somebody works on it

**Comment by WimSpee** [comment]
There now is a ` bcftools +split ` plugin which can split a multi-sample VCF file to many single sample VCF files. Did not yet try it.
http://samtools.github.io/bcftools/bcftools.html#plugin

This would help in implementing option 2 described in the issue opening.

**Comment by naumenko-sa** [comment]
Have you tried installing bcftools plugins with conda?
https://github.com/bioconda/bioconda-recipes/issues/23993

**Comment by WimSpee** [comment]
No, I have not tried.  Also I don't know if I ever used the pugins from the bcftools that is installed by bcbio. 

I have used the plugins form the bcftools downloaded and installed from github as described here
http://www.htslib.org/download/

Not sure if the bcftools pugins are compatible with the current installation setup via conda. 

By the way, when doing a joint analysis from existing GVCF files and a callable bed file, the bcftools stats step is skipped. 
https://github.com/bcbio/bcbio-nextgen/issues/3380

This solves the performance issue, but the bcftools stats are nice to have. 
So we are now looking at implementing bcftools stats as a custom internal  step after that bcbio joint analysis has finished. 

Still would be nice of course if the bcftools stats would be faster as supported by bcbio .

I hope this information is useful.
